pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/rebalance/DefaultRebalanceSegmentStrategy.java

```diff
@@ -0,0 +1,306 @@
+      int numPartitions = getKafkaPartitionCount(tableConfig);
+
+      Map<String, PartitionAssignment> consumingPartitionAssignment =
+          streamPartitionAssignmentGenerator.generatePartitionAssignment(tableConfig, numPartitions, consumingInstances,
+              Lists.newArrayList(tableNameWithType));
+      partitionAssignment = consumingPartitionAssignment.get(tableNameWithType);
+
+      boolean dryRun =
+          Boolean.valueOf(rebalanceUserConfig.getConfig(RebalanceUserConfigProperties.DRYRUN, DEFAULT_DRY_RUN));
+      if (!dryRun) {
+        LOGGER.info("Dry run. Skip writing stream partition assignment to property store");
+        streamPartitionAssignmentGenerator.writeKafkaPartitionAssignment(consumingPartitionAssignment);
+      }
+    }
+    return partitionAssignment;
+  }
+
+  @Override
+  public IdealState rebalanceIdealState(IdealState idealState, TableConfig tableConfig,
+      RebalanceUserConfig rebalanceUserConfig, PartitionAssignment newPartitionAssignment) {
+
+    String tableNameWithType = tableConfig.getTableName();
+    CommonConstants.Helix.TableType tableType = tableConfig.getTableType();
+
+    // if realtime, rebalance consuming segments
+    if (tableType.equals(CommonConstants.Helix.TableType.REALTIME)) {
+      boolean rebalanceConsuming = Boolean.valueOf(
+          rebalanceUserConfig.getConfig(RebalanceUserConfigProperties.REBALANCE_CONSUMING,
+              DEFAULT_REBALANCE_CONSUMING));
+      if (rebalanceConsuming) {
+        rebalanceConsumingSegments(idealState, newPartitionAssignment);
+      }
+    }
+    // rebalance serving segments
+    int targetNumReplicas;
+    if (tableType.equals(CommonConstants.Helix.TableType.REALTIME)) {
+      String replicasString = tableConfig.getValidationConfig().getReplicasPerPartition();
+      try {
+        targetNumReplicas = Integer.parseInt(replicasString);
+      } catch (Exception e) {
+        throw new RuntimeException("Invalid value for replicasPerPartition:'" + replicasString + "'", e);
+      }
+    } else {
+      targetNumReplicas = Integer.parseInt(tableConfig.getValidationConfig().getReplication());
+    }
+    rebalanceServing(idealState, tableConfig, targetNumReplicas);
+    boolean dryRun =
+        Boolean.valueOf(rebalanceUserConfig.getConfig(RebalanceUserConfigProperties.DRYRUN, DEFAULT_DRY_RUN));
+    if (!dryRun) {
+      updateIdealStateWithNewSegmentMapping(tableNameWithType, targetNumReplicas,
+          idealState.getRecord().getMapFields());
+    }
+    return idealState;
+  }
+
+  /**
+   * Helper method to update idealstate with the new segment assignment
+   *
+   * @param tableNameWithType Table name with type
+   * @param numReplica The number of replica
+   * @param segmentAssignmentMapping Segment assignment mapping
+   */
+  private void updateIdealStateWithNewSegmentMapping(final String tableNameWithType, final int numReplica,
```

Comment:
I think this function can be shared across the `RebalanceSegmentStrategy`. One way to achieve is to make an abstract class `BaseRebalanceSegmentStrategy` and make all the implementation extends that. Another way is to change the interface not to include the update to zk and pull this function up to `PinotHelixResourceManager`. I personally prefer to the latter.

Reply:
I had thought of putting it in abstract base class too, but was confused between which approach to take. I like the approach of abstract class better, because that way all responsibility of writing (partition assignment and ideal state)  is with the implementation and not with the caller. If we move updateIdealState to HelixResourceManager, the writing of partition assignment would still happen in the implementation, whereas the writing of ideal state will happen outside. If we go with approach 2, should we also be pulling out writePartitionAssignment to PinotHelixManager? And in the future if another strategy wants to write something else, should that be moved too?